src/Interpreters/QueryMetricLog.cpp

```diff
@@ -0,0 +1,206 @@
+                "Query ID."});
+    result.add({"hostname",
+                std::make_shared<DataTypeLowCardinality>(std::make_shared<DataTypeString>()),
+                parseQuery(codec_parser, "(ZSTD(1))", 0, DBMS_DEFAULT_MAX_PARSER_DEPTH, DBMS_DEFAULT_MAX_PARSER_BACKTRACKS),
+                "Hostname of the server executing the query."});
+    result.add({"event_date",
+                std::make_shared<DataTypeDate>(),
+                parseQuery(codec_parser, "(Delta(2), ZSTD(1))", 0, DBMS_DEFAULT_MAX_PARSER_DEPTH, DBMS_DEFAULT_MAX_PARSER_BACKTRACKS),
+                "Event date."});
+    result.add({"event_time",
+                std::make_shared<DataTypeDateTime>(),
+                parseQuery(codec_parser, "(Delta(4), ZSTD(1))", 0, DBMS_DEFAULT_MAX_PARSER_DEPTH, DBMS_DEFAULT_MAX_PARSER_BACKTRACKS),
+                "Event time."});
+    result.add({"event_time_microseconds",
+                std::make_shared<DataTypeDateTime64>(6),
+                parseQuery(codec_parser, "(Delta(4), ZSTD(1))", 0, DBMS_DEFAULT_MAX_PARSER_DEPTH, DBMS_DEFAULT_MAX_PARSER_BACKTRACKS),
+                "Event time with microseconds resolution."});
+    result.add({"memory_usage",
+                std::make_shared<DataTypeUInt64>(),
+                "Amount of RAM the query uses. It might not include some types of dedicated memory."});
+    result.add({"peak_memory_usage",
+                std::make_shared<DataTypeUInt64>(),
+                "Maximum amount of RAM the query used."});
+
+    for (size_t i = 0, end = ProfileEvents::end(); i < end; ++i)
+    {
+        auto name = fmt::format("ProfileEvent_{}", ProfileEvents::getName(ProfileEvents::Event(i)));
+        const auto * comment = ProfileEvents::getDocumentation(ProfileEvents::Event(i));
+        result.add({std::move(name), std::make_shared<DataTypeUInt64>(), comment});
+    }
+
+    return result;
+}
+
+void QueryMetricLogElement::appendToBlock(MutableColumns & columns) const
+{
+    size_t column_idx = 0;
+
+    columns[column_idx++]->insert(query_id);
+    columns[column_idx++]->insert(getFQDNOrHostName());
+    columns[column_idx++]->insert(DateLUT::instance().toDayNum(event_time).toUnderType());
+    columns[column_idx++]->insert(event_time);
+    columns[column_idx++]->insert(event_time_microseconds);
+    columns[column_idx++]->insert(memory_usage);
+    columns[column_idx++]->insert(peak_memory_usage);
+
+    for (size_t i = 0, end = ProfileEvents::end(); i < end; ++i)
+        columns[column_idx++]->insert(profile_events[i]);
+}
+
+void QueryMetricLog::shutdown()
+{
+    Base::shutdown();
+}
+
+void QueryMetricLog::startQuery(const String & query_id, TimePoint query_start_time, UInt64 interval_milliseconds)
+{
+    QueryMetricLogStatus status;
+    status.interval_milliseconds = interval_milliseconds;
+    status.next_collect_time = query_start_time + std::chrono::milliseconds(interval_milliseconds);
+
+    const auto & profile_events = CurrentThread::getProfileEvents();
```

Comment:
>It seems that whatever events where triggered before calling QueryMetricLog::startQuery won't be part of any log (not even the first for the query), which is kind of unexpected

That's right. `query_metric_log` works exactly the same as `metric_log`, but for particular queries. This means that we store the state at the beginning and then we report periodically the difference between the new state and the previous one. So, as your comment suggests, if there's a single profile event that happens right before this call that is never modified during the execution of the query, it won't be reported in any event of `query_metric_log`. `QueryMetricLog::startQuery` is already called on the early stage of the query execution, but there's stuff that may happen before it for sure. For getting those sort of edge cases we'll need to go to the global `metric_log`.

The best I can think of is moving the call to `QueryMetricLog::startQuery` as early as possible in an attempt to catch every profile event triggered by the query, even at the early stage.

Reply:
But why ignore it at all for the first instance? I assume profileEvents are 0 at the very start of the query, so they should be there in the first row that dumps them to system.query_metric_log even if they happen before `QueryMetricLog::startQuery`

Reply:
>I assume profileEvents are 0 at the very start of the query

Yep, I double checked just in case and it's like that: https://github.com/ClickHouse/ClickHouse/blob/4baa30c06d7e7996c548b85120044ca691ff2326/src/Interpreters/ThreadStatusExt.cpp#L383-L385

>But why ignore it at all for the first instance?

I see your point now. It's a decision I made in an attempt to save only the minimum amount of info because I assumed the first row would be of little value. Regarding this, these metrics are only saved once the query exceeds the collection interval (1s by default). In any case, the same way I ended up adding the finish row because I think it's very useful, I can add the start row to catch that case. Will do :+1.

Reply:
> I see your point now. It's a decision I made in an attempt to save only the minimum amount of info because I assumed the first row would be of little value.

I'm not talking about saving the values at the start of the query (one extra row), but about not discarding the profileEvents that happened before start of query (typically from planning, which can be expensive) so they are dumped when the first row is stored (1 second in, by default). The storage usage should be pretty much the same

Reply:
Done at [`4223bb9` (#66532)](https://github.com/ClickHouse/ClickHouse/pull/66532/commits/4223bb95a50fde3155ae3d5f02eaf0a95a2e3ec4)

Reply:
> I'm not talking about saving the values at the start of the query (one extra row), but about not discarding the profileEvents that happened before start of query (typically from planning, which can be expensive) so they are dumped when the first row is stored (1 second in, by default). The storage usage should be pretty much the same

Damn, I misunderstood and implemented the other thing. I reverted it and finally did what you were proposing at [`795abfc` (#66532)](https://github.com/ClickHouse/ClickHouse/pull/66532/commits/795abfcb0a78202967d915bcac707ca9cb6b340c)